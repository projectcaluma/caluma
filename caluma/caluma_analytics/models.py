from django.contrib.postgres.fields import ArrayField
from django.db import models
from django.db.models.constraints import UniqueConstraint
from localized_fields.fields import LocalizedField

from caluma.caluma_core.models import SlugModel, UUIDModel

from .simple_table import BaseStartingObject


class AnalyticsTable(SlugModel):
    STARTING_OBJECT_CHOICES = BaseStartingObject.as_choices()
    TYPE_EXTRACTION = "type_extraction"
    TYPE_PIVOT = "type_pivot"

    TABLE_TYPE_CHOICES = (
        (TYPE_EXTRACTION, "Data extraction"),
        (TYPE_PIVOT, "Pivot table"),
    )

    meta = models.JSONField(default=dict)

    table_type = models.CharField(max_length=250, choices=TABLE_TYPE_CHOICES, null=True)

    base_table = models.ForeignKey(
        "self",
        on_delete=models.RESTRICT,
        related_name="following_analytics_tables",
        null=True,
        default=None,
    )

    disable_visibilities = models.BooleanField(default=False)
    name = LocalizedField(blank=False, null=False, required=False)
    starting_object = models.CharField(
        max_length=250, choices=STARTING_OBJECT_CHOICES, null=True
    )

    def get_starting_object(self, info):
        return BaseStartingObject.get_object(self.starting_object, info)

    def __repr__(self):
        return f"AnalyticsTable<{self.slug}>"


class AnalyticsField(UUIDModel):
    """Analytics field.

    May represent an aggregate (pivot) field on a pivot table, or
    an extraction field on an extraction table.

    The data_source, in extraction fields, references a field
    listed from the available fields that the starting object of
    the table provides. In pivot tables, the data_source references
    a field of the base table, by its alias .

    The function (if it's a pivot field) defines how the data of the
    specified source field is aggregated. While most of the functions
    are obvious by their name, the `FUNCTION_GROUP` is special: Those
    fields will not be aggregated, but the other aggregates will be
    grouped by this value.
    For example, if you define the following fields:
        - foo SUM
        - bar MIN
        - baz GROUP
    The output will contain one row for each unique value of `baz`.
    The sums and minimums calculated will all be in relation to rows with
    the same `baz` value.
    """

    FUNCTION_GROUP = "group"
    FUNCTION_SUM = "sum"
    FUNCTION_AVERAGE = "avg"
    FUNCTION_MAX = "max"
    FUNCTION_MIN = "min"

    FUNCTION_CHOICES = [
        (FUNCTION_GROUP, "Group by"),
        (FUNCTION_SUM, "Sum"),
        (FUNCTION_AVERAGE, "Average"),
        (FUNCTION_MAX, "Max"),
        (FUNCTION_MIN, "Min"),
    ]

    alias = models.CharField(max_length=100)
    meta = models.JSONField(default=dict)

    data_source = models.TextField()
    table = models.ForeignKey(
        AnalyticsTable, on_delete=models.CASCADE, related_name="fields"
    )
    filters = ArrayField(models.TextField(), blank=True, null=True, default=list)
    function = models.CharField(
        max_length=20, choices=FUNCTION_CHOICES, null=True, default=None
    )

    def __repr__(self):
        return f"AnalyticsField<{self.table.slug}.{self.alias}>"

    class Meta:
        constraints = [
            UniqueConstraint(
                name="unique_data_source", fields=["table", "data_source"]
            ),
            UniqueConstraint(name="unique_alias", fields=["table", "alias"]),
        ]
