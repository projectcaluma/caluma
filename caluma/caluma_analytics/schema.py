import graphene
from django.db.models.query_utils import DeferredAttribute
from graphene import ConnectionField, String, relay
from graphene.types import ObjectType, generic
from graphene_django.rest_framework import serializer_converter

from ..caluma_core.filters import (
    CollectionFilterSetFactory,
    DjangoFilterSetConnectionField,
)
from ..caluma_core.mutation import Mutation, UserDefinedPrimaryKeyMixin
from ..caluma_core.types import CountableConnectionBase, DjangoObjectType
from . import filters, models, pivot_table, serializers, simple_table


class AvailableField(ObjectType):
    """Available fields show users what can be selected in an analysis.

    The main identifier is the source path, but for display purposes,
    a label (field at current position) and a full_label (including
    parent fields' labels) is available.

    Frontends should query sub-fields (via prefix/depth) if is_leaf is
    False. Some fields can be non-leafs as well as values, such as
    dates: Dates can be extracted "as is", or we can extract a
    date part (such as year, quarter, ...) from it.
    """

    label = String()
    full_label = String()
    source_path = String()
    is_leaf = graphene.Boolean()
    is_value = graphene.Boolean()

    class Meta:
        interfaces = (relay.Node,)
        connection_class = CountableConnectionBase


class AvailableFieldConnection(CountableConnectionBase):
    class Meta:
        node = AvailableField


class AnalyticsCell(ObjectType):
    """A cell represents one value in the analytics output."""

    alias = String()
    value = String(required=False)


class AnalyticsRow(graphene.Connection):
    class Meta:
        node = AnalyticsCell


class AnalyticsTableContent(graphene.Connection):
    class Meta:
        node = AnalyticsRow


class AnalyticsOutput(ObjectType):
    records = ConnectionField(AnalyticsTableContent)

    @staticmethod
    def resolve_records(table, info, *args, **kwargs):
        rows = [
            AnalyticsRow(
                edges=[
                    {"node": {"alias": alias, "value": val}}
                    for alias, val in row.items()
                ]
            )
            for row in table.get_records()
        ]
        return rows


def enum_type_from_field(name, field, description=None, register_to_field=None):
    if isinstance(field, DeferredAttribute):
        field = field.field  # pragma: no cover

    the_type = graphene.Enum(
        name, [(key.upper(), key) for key, _ in field.choices], description=description
    )

    if register_to_field:
        serializer_converter.get_graphene_type_from_serializer_field.register(
            register_to_field, lambda field: the_type
        )

    return the_type


StartingObject = enum_type_from_field(
    "StartingObject",
    models.AnalyticsTable.starting_object,
    register_to_field=serializers.StartingObjectField,
)


AggregateFunction = enum_type_from_field(
    "AggregateFunction",
    models.AnalyticsField.function,
    description="Aggregate function for pivot table",
    register_to_field=serializers.AggregateFunctionField,
)


TableType = graphene.Enum(
    "TableType",
    [(key.upper(), key) for key, _ in models.AnalyticsTable.TABLE_TYPE_CHOICES],
)

serializer_converter.get_graphene_type_from_serializer_field.register(
    serializers.TableTypeField, lambda field: TableType
)


class AnalyticsTable(DjangoObjectType):
    available_fields = ConnectionField(
        AvailableFieldConnection,
        prefix=String(required=False),
        depth=graphene.Int(required=False),
    )
    result_data = graphene.Field(AnalyticsOutput)
    starting_object = StartingObject(required=False)
    table_type = TableType(required=False)

    @staticmethod
    def resolve_available_fields(obj, info, prefix=None, depth=None, **kwargs):
        start = obj.get_starting_object(info)

        depth = depth if depth and depth > 0 else 1
        prefix = prefix.split(".") if prefix else []

        return sorted(
            [
                {
                    "id": ".".join(field.source_path()),
                    "label": field.label,
                    "full_label": field.full_label(),
                    "source_path": ".".join(field.source_path()),
                    "is_leaf": field.is_leaf(),
                    "is_value": field.is_value(),
                }
                for path, field in start.get_fields(prefix, depth).items()
            ],
            key=lambda field: field["id"],
        )

    result_data = graphene.Field(AnalyticsOutput)

    @staticmethod
    def resolve_result_data(obj, info):
        if obj.table_type == obj.TYPE_PIVOT:
            return pivot_table.PivotTable(obj)
        elif obj.table_type == obj.TYPE_EXTRACTION:
            return simple_table.SimpleTable(obj)

    class Meta:
        model = models.AnalyticsTable
        interfaces = (relay.Node,)
        connection_class = CountableConnectionBase
        fields = [
            "slug",
            "meta",
            "created_by_group",
            "created_by_user",
            "modified_by_group",
            "modified_by_user",
            "created_at",
            "modified_at",
            "disable_visibilities",
            "available_fields",
            "result_data",
            "fields",
            "name",
            "base_table",
            "table_type",
            "starting_object",
        ]


class AnalyticsField(DjangoObjectType):
    meta = generic.GenericScalar()
    filters = graphene.List(String, required=False)
    function = AggregateFunction(required=False)

    @classmethod
    def get_queryset(cls, queryset, info):
        return super().get_queryset(queryset, info).order_by("-created_at")

    class Meta:
        model = models.AnalyticsField
        interfaces = (relay.Node,)
        connection_class = CountableConnectionBase


class SaveAnalyticsTable(UserDefinedPrimaryKeyMixin, Mutation):
    class Meta:
        serializer_class = serializers.SaveAnalyticsTableSerializer
        model_operations = ["create", "update"]


class RemoveAnalyticsTable(Mutation):
    class Meta:
        serializer_class = serializers.RemoveAnalyticsTableSerializer
        lookup_input_kwarg = "slug"
        model_operations = ["update"]


class SaveAnalyticsField(Mutation):
    class Meta:
        serializer_class = serializers.SaveAnalyticsFieldSerializer
        model_operations = ["create", "update"]


class RemoveAnalyticsField(Mutation):
    class Meta:
        serializer_class = serializers.RemoveAnalyticsFieldSerializer
        lookup_input_kwarg = "id"
        model_operations = ["update"]


class Mutation:
    save_analytics_table = SaveAnalyticsTable().Field()
    remove_analytics_table = RemoveAnalyticsTable().Field()

    save_analytics_field = SaveAnalyticsField().Field()
    remove_analytics_field = RemoveAnalyticsField().Field()


class Query:
    all_analytics_tables = DjangoFilterSetConnectionField(
        AnalyticsTable._meta.connection,
        filterset_class=CollectionFilterSetFactory(
            filters.AnalyticsTableFilterSet,
            orderset_class=filters.AnalyticsTableOrderSet,
        ),
    )
    analytics_table = graphene.Field(AnalyticsTable, slug=String(required=True))

    all_analytics_fields = DjangoFilterSetConnectionField(
        AnalyticsField._meta.connection,
        filterset_class=CollectionFilterSetFactory(filters.AnalyticsFieldFilterSet),
    )

    @staticmethod
    def resolve_analytics_table(_, info, slug):
        return models.AnalyticsTable.objects.get(pk=slug)
