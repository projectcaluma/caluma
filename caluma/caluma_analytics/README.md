# Caluma Analytics

This documentation is aimed at programmers working on the analytics module.
To read about how to use the analytics module, refer to the user docs linked
from the main README.

## General workings / Overall concept

The analytics module's functionality is guided by the "Pivot Table"
feature found in Microsoft Excel and LibreOffice Calc. There are two
steps for a full analysis view:

1) Generate a table of data to be analyzed
2) Apply the pivot table mechanism on selected fields

The queries are  stored as objects in the database, and can be queried
via GraphQL, or via commandline. This way, the module splits into a
design time part and a runtime part, the same as the rest of Caluma.

The two steps mentioned above are reflected in the data model as well.
You can create a "simple table" that doesn't do any aggregation at all.
This table is then used as input to a "pivot table" which runs sums, counts,
averages and so on.

### Simple tables

Simple tables are used to collect data from the production environment.
A simple table consists of one *Starting point*, as well as multiple *fields*.
The starting point defines the entity that will be represented by each row
of the table output. For instance, when using the "Cases" starting point,
each row in the simple table output will represent one case.

Then, there are multiple fields. The fields represent a path to the
information of interest, relative to the starting point. The internal
representation of this field defines how to access the given value.
Each field has a label, and a path in dot-notation, for example
`case.document.answers.total_price`.

### Simple table fields

For example, the `AttributeField` defines how to access an attribute of
a given database table. It could be a `created_at` date field, for example,
or any other field.

When the required data is in another table than the main table, we use a
`JoinField`. This field knows how to SQL `JOIN` the required table.

Each field may have a "parent" field. When generating the SQL for the analysis,
this is used to add it to the correct subquery.

Due to the fact that a simple table row always contains exactly one row
of output per starting object entity, special care must be taken when
extracting data form a 1:N connection (such as work item data from a case):
Since we cannot get multiple work items for the same case, the field
must ensure that only one row of data is returned per "parent-row". This
is done by being explicit in what we select: The workitem path will not only
join the work items, but it needs to take either the first, or the last one,
depending on the user's needs. `workitems[task=ordering,first-by-created]`
and `workitems[task=ordering,last-by-created]` would be two examples
of this.

### Filtering

Each field, when specified on a simple table, may have a filter
associated.  For simplicity's sake, and because we don't have the
full filtering infrastructure of Graphene at our disposal, we only
provide a simple list of values as filtering mechanism: The user can,
at design-time, provide a list of values. Only rows where the value of
the given field matches one of the values will be shown in the output.

## Query generation

When generating queries, we unfortunately cannot use Django's ORM, as the
queries we're generating are too complex, and the output does not match
the ORM's model anyway, so it is of little actual use.

The basic query structure is that we're heavily relying on subqueries, and
join them in the usual manner. When each field is added to the query structure,
we iterate through the field's parents top-down and reuse already-joined tables.
This way, we don't have to join the same data multiple times.

Each `JoinField` defines an inner and outer ref, along the tables and other
conditions that might apply. Those references are used to ensure that the
inner data corresponds to the outer data.
Each subquery is also aliased to a unique name. Subqueries are only reused if
they actually refer to the exact same set of data.

Each query selects all the fields from the inner subquery explicitly by name.
This way, in the outermost query, we have access to all the data from any
subquery. Filtering can thus happen in the outermost query: We cannot do the
filtering in inner queries for the fact that we still want a row in the output
even if one of the fields requested is not available (or NULL). If we did the
filtering "inside", we'd need to use `INNER JOIN`s and would lose this ability.

What if we need to access multiple answers from a form, for example? In this
case, we would generate two distinct subqueries that each retrieves the data
from the "answers" table. The JOIN conditions would be the same, but the inner
condition would differ (namely referring another question id).

### Visibility

Even though we're not using Django's ORM, we still need to respect the
visibilities defined in the Caluma configuration. This is done by fetching a
"plain" queryset from a given model, then filtering it through the visibility
layer. From the resulting queryset, the SQL and parameters are extracted and
used as a basis for the analysis data.


## Code structure

There are a few modules that contain distinct aspects of the code:

* `simple_table.py`: Simple data extraction table, and field definition
* `sql.py`: Generates SQL from the table definitions
* `visibility.py`: Gives us the visibility-filtered querysets
