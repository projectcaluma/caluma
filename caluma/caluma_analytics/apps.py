from django.apps import AppConfig


class DefaultConfig(AppConfig):
    name = "caluma.caluma_analytics"
