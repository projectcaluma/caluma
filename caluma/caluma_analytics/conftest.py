import pytest

from caluma.caluma_workflow import models as workflow_models

from . import models


@pytest.fixture
def example_analytics(analytics_table, form_and_document, settings):
    settings.META_FIELDS = ["foo"]
    analytics_table.table_type = models.AnalyticsTable.TYPE_EXTRACTION

    analytics_table.fields.create(data_source="created_at", alias="created_at")
    analytics_table.fields.create(data_source="created_at.quarter", alias="quarter")
    analytics_table.fields.create(data_source="meta.foo", alias="foo")
    analytics_table.fields.create(data_source="status", alias="statuuuuuus")
    analytics_table.fields.create(
        data_source="document[top_form].top_question", alias="from_the_doc"
    )
    return analytics_table


@pytest.fixture
def analytics_cases(form_and_document, case_factory, at_date):
    """Create a number of cases with documents to be used in testing.

    There are multiple cases with differing statuses and differing
    created_at values:
        * 3x RUNNING, created_at ranging from 2022-02-01 to 2022-01-03
        * 1x COMPLETED, created_at 2022-02-04
        * 1x SUSPENDED, created_at 2022-02-05
    """
    _f, document, _q, _a = form_and_document(use_subform=True)
    statuses = [
        # multiple cases with same status to test aggregates
        workflow_models.Case.STATUS_RUNNING,  # created_at 2022-02-01
        workflow_models.Case.STATUS_RUNNING,  # created_at 2022-02-02
        workflow_models.Case.STATUS_RUNNING,  # created_at 2022-02-03
        workflow_models.Case.STATUS_COMPLETED,  # created_at 2022-02-04
        workflow_models.Case.STATUS_SUSPENDED,  # created_at 2022-02-05
    ]

    # create each case with another modified_at date

    return [
        # use at_date to create the cases at the correct date.
        at_date(
            f"2022-02-{day:02}",
            lambda: case_factory(
                meta={"foo": "bar"},
                status=status,
                document=document.copy(),
            ),
        )
        for day, status in enumerate(statuses, start=1)
    ]


@pytest.fixture
def example_pivot_table(example_analytics, analytics_table_factory):
    pivot_table = analytics_table_factory(
        table_type=models.AnalyticsTable.TYPE_PIVOT, base_table=example_analytics
    )

    pivot_table.fields.create(
        alias="status",
        data_source="statuuuuuus",
        function=models.AnalyticsField.FUNCTION_GROUP,
    )

    pivot_table.fields.create(
        alias="last_created",
        data_source="created_at",
        function=models.AnalyticsField.FUNCTION_MAX,
    )
    return pivot_table
