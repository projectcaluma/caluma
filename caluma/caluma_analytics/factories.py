from factory import Faker, LazyAttribute, Maybe, SubFactory

from ..caluma_core.factories import DjangoModelFactory
from . import models


class AnalyticsTableFactory(DjangoModelFactory):
    slug = Faker("slug")
    name = Faker("multilang", faker_provider="name")
    starting_object = Faker(
        "word",
        ext_word_list=[i for i, _ in models.AnalyticsTable.STARTING_OBJECT_CHOICES],
    )
    table_type = Faker(
        "word",
        ext_word_list=[i for i, _ in models.AnalyticsTable.TABLE_TYPE_CHOICES],
    )
    base_table = None

    meta = {}

    class Meta:
        model = models.AnalyticsTable


class AnalyticsFieldFactory(DjangoModelFactory):
    alias = Faker("slug")
    meta = {}
    data_source = Faker("slug")
    table = SubFactory(AnalyticsTableFactory)

    function = Maybe(
        "is_pivot",
        yes_declaration=Faker(
            "word",
            ext_word_list=[i for i, _ in models.AnalyticsField.FUNCTION_CHOICES],
        ),
        no_declaration=None,
    )

    class Meta:
        model = models.AnalyticsField

    class Params:
        is_pivot = LazyAttribute(
            lambda f: f.table.table_type == models.AnalyticsTable.TYPE_PIVOT
        )
