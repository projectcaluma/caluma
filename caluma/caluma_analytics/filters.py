from django_filters.rest_framework import FilterSet

from ..caluma_core.filters import MetaFilterSet, SearchFilter, SlugMultipleChoiceFilter
from ..caluma_core.ordering import AttributeOrderingFactory, MetaFieldOrdering
from . import models


class AnalyticsTableFilterSet(MetaFilterSet):
    search = SearchFilter(fields=("slug", "name"))
    slugs = SlugMultipleChoiceFilter(field_name="slug")

    class Meta:
        model = models.AnalyticsTable
        fields = (
            "slug",
            "name",
        )


class AnalyticsFieldFilterSet(MetaFilterSet):
    search = SearchFilter(
        fields=(
            "alias",
            "data_source",
            "function",
            "table__name",
            "source_field__alias",
        )
    )
    slugs = SlugMultipleChoiceFilter(field_name="slug")

    class Meta:
        model = models.AnalyticsField
        fields = (
            "alias",
            "table",
        )


class AnalyticsTableOrderSet(FilterSet):
    meta = MetaFieldOrdering()
    attribute = AttributeOrderingFactory(
        models.AnalyticsTable,
        fields=[
            "created_at",
            "modified_at",
            "created_by_user",
            "created_by_group",
            "modified_by_user",
            "modified_by_group",
            "slug",
            "name",
        ],
    )

    class Meta:
        model = models.AnalyticsTable
        fields = ("meta", "attribute")
