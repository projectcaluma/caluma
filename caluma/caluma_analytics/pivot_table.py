from __future__ import annotations  # self-referencing annotations

from functools import cached_property

from django.db import connection
from psycopg2.extras import DictCursor

from . import simple_table, sql


class PivotTable:
    class _AnonymousInfo:
        """
        Pseudo GraphQL info object.

        Used for commandline access, so the commandline version
        can also invoke the visibilities.
        """

        context = {}

    def __init__(self, table, info=_AnonymousInfo):
        self.info = info

        self.table = table
        self.last_query = None
        self.last_query_params = None

    @cached_property
    def _fields(self):
        return list(self.table.fields.all())

    def get_sql_and_params(self):
        """Return a list of records as specified in the given table config."""

        group_by = []
        base_table = simple_table.SimpleTable(self.table.base_table)

        # We replace the selects from the analytics query with
        # the aggregate function calls as required, but leave the
        # rest intact. This way, we won't need to build a subquery
        # around the whole thing and keep it relatively (!) simple
        aggregate_query = base_table.get_query_object()

        selects_by_alias = {
            alias: fragment for fragment, alias in aggregate_query.select
        }

        new_selects = []

        for field in self._fields:
            key = f"analytics_result_{field.data_source}"

            old_field = selects_by_alias[key]
            new_alias = f"analytics_result_{field.alias}"

            if field.function == field.FUNCTION_GROUP:
                group_by.append(old_field)
                new_selects.append((old_field, new_alias))
            else:
                new_selects.append((f"{field.function}({old_field})", new_alias))

        aggregate_query.select = new_selects
        aggregate_query.group_by.extend(group_by)
        aggregate_query.select_direct_only = True

        sql_query, params, _ = sql.QueryRender(aggregate_query).as_sql(alias=None)

        return sql_query, params

    def get_records(self):

        # TODO: this is exactly the same as SimpleTable.get_records(). Can we share code?
        sql_query, params = self.get_sql_and_params()

        with connection.connection.cursor(cursor_factory=DictCursor) as cursor:
            cursor.execute(sql_query, params)
            data = cursor.fetchall()

            return [
                {
                    # TODO: do we need a parse_value() equivalent here?
                    field.alias: row[f"analytics_result_{field.alias}"]
                    for field in self._fields
                }
                for row in data
            ]

    def _sql_alias(self, user_alias):  # pragma: no cover
        return f"analytics_pivot_{user_alias}"
