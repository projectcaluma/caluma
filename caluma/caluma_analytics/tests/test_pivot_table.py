import pytest

from caluma.caluma_analytics.pivot_table import PivotTable


def _table_output_to_rows(output):
    """Convert GQL output (list of list of gql dicts) into simple list of dicts."""
    return [
        {col["node"]["alias"]: col["node"]["value"] for col in rec["node"]["edges"]}
        for rec in output
    ]


@pytest.mark.freeze_time("2021-10-10")
@pytest.mark.parametrize("analytics_table__starting_object", ["cases"])
def test_run_analytics_direct(db, snapshot, example_pivot_table, analytics_cases):
    """Test basic analytics run on simple table.

    * Test that we get *some* results
    * Tests date part extraction
    * Test Meta field extraction
    """

    table = PivotTable(example_pivot_table)

    result = table.get_records()

    assert len(result) == 3

    snapshot.assert_match(sorted(result, key=lambda r: r["status"]))


@pytest.mark.parametrize("analytics_table__starting_object", ["cases"])
@pytest.mark.freeze_time("2021-10-10")
@pytest.mark.parametrize("delete_case", [True, False])
@pytest.mark.parametrize("case__meta", [{"foo": "bar"}])
def test_run_analytics_gql(
    db,
    schema_executor,
    example_pivot_table,
    case,
    delete_case,
    analytics_cases,
):
    """Test basic analytics run on simple table.

    * Test that we get *some* results
    * Tests date part extraction
    * Test Meta field extraction
    """
    query = """
        query run ($input: String!) {
          analyticsTable(slug: $input) {
              resultData {
                records {
                  edges {
                    node {
                      edges {
                        node {
                          alias
                          value
                        }
                      }
                    }
                  }
                }
                summary {
                  edges {
                    node {
                      alias
                      value
                    }
                  }
                }
             }
          }
       }
    """

    if delete_case:
        # Test empty output
        case.delete()

    result = schema_executor(query, variable_values={"input": example_pivot_table.pk})
    assert not result.errors

    if not delete_case:
        # some explicit sanity checks
        records = result.data["analyticsTable"]["resultData"]["records"]["edges"]
        assert len(records) == 3
        data = _table_output_to_rows(records)
        created_by_status = {row["status"]: row["last_created"] for row in data}

        assert created_by_status == {
            "running": "2022-02-05 00:00:00+00:00",
            "completed": "2022-02-04 00:00:00+00:00",
            "suspended": "2022-02-05 00:00:00+00:00",
        }

        summary = result.data["analyticsTable"]["resultData"]["summary"]["edges"]

        summary_dict = {col["node"]["alias"]: col["node"]["value"] for col in summary}
        assert summary_dict == {
            "status": "0",
            "last_created": "0",
            "quarter": "0",
            "sub_question_sumsumsum": "17448.0",
        }
