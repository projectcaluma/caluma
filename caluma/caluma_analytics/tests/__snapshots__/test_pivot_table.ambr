# name: test_run_analytics_direct
  <class 'list'> [
    <class 'dict'> {
      'last_created': datetime.datetime(2022, 2, 4, 0, 0, tzinfo=datetime.timezone.utc),
      'status': 'completed',
    },
    <class 'dict'> {
      'last_created': datetime.datetime(2022, 2, 3, 0, 0, tzinfo=datetime.timezone.utc),
      'status': 'running',
    },
    <class 'dict'> {
      'last_created': datetime.datetime(2022, 2, 5, 0, 0, tzinfo=datetime.timezone.utc),
      'status': 'suspended',
    },
  ]
---
